### PR TITLE
Deactivate user in WIT and Tenant

### PR DESCRIPTION
**Pull request by xcoulon**

### CODE REVIEW

**Comment by xcoulon** [comment]
requires #789 to be merged, first.

**Comment by alexeykazakov** [comment]
Is it still WIP?

**Comment by xcoulon** [comment]
> Is it still WIP?

yes, until #789 is merged, which will happen very soon. I'll ping you again once it's ready for review

**Comment by xcoulon** [comment]
@alexeykazakov it's ready for your reviewing pleasure, now ;)

**Comment by xcoulon** [comment]
[test]

File: authentication/account/service/user.go [hunk 1]
```diff
@@ -145,8 +145,27 @@ func (s *userServiceImpl) DeactivateUser(ctx context.Context, username string) (
 		}
 
 		// revoke all user's tokens
-		return s.Services().TokenService().SetStatusForAllIdentityTokens(ctx, identity.ID, token.TOKEN_STATUS_REVOKED)
+		err = s.Services().TokenService().SetStatusForAllIdentityTokens(ctx, identity.ID, token.TOKEN_STATUS_REVOKED)
+		if err != nil {
+			return err
+		}
+
+		return nil
```
Comment (alexeykazakov) [inline authentication/account/service/user.go]: You don't need to change this at all, do you? 
`return s.Services().TokenService().SetStatusForAllIdentityTokens(ctx, identity.ID, token.TOKEN_STATUS_REVOKED)` works fine here.

File: authentication/account/service/user.go [hunk 2]
```diff
@@ -145,8 +145,27 @@ func (s *userServiceImpl) DeactivateUser(ctx context.Context, username string) (
...
 	})
+	if err != nil {
+		return nil, err
+	}
+
+	// call WIT and Tenant to deactivate the user there as well,
+	// using `auth` SA token here, not the request context's token
+	err = s.Services().WITService().DeleteUser(ctx, identity.ID.String())
+	if err != nil {
+		return nil, err
```
Comment (alexeykazakov) [inline authentication/account/service/user.go]: We are about to remote WIT :) Let's not make hard dependencies to it. Just log the error and proceed maybe?

File: authentication/account/service/user_blackbox_test.go [hunk 3]
```diff
@@ -75,8 +82,26 @@ func (s *userServiceBlackboxTestSuite) TestDeactivate() {
...
 		token3 := s.Graph.CreateToken(userToStayIntact)
 		token4 := s.Graph.CreateToken(userToStayIntact)
+
+		defer gock.OffAll()
+		// call to WIT Service
+		witCallsCounter := 0
+		gock.New("http://localhost:8080").
+			Delete(fmt.Sprintf("/api/users/username/%s", userToDeactivate.IdentityID().String())).
+			MatchHeader("Authorization", "Bearer "+saToken).
+			// MatchHeader("X-Request-Id", reqID).
```
Comment (alexeykazakov) [inline authentication/account/service/user_blackbox_test.go]: Why commented it out?

File: authentication/account/service/user_blackbox_test.go [hunk 4]
```diff
@@ -75,8 +82,26 @@ func (s *userServiceBlackboxTestSuite) TestDeactivate() {
...
+			SetMatcher(gocksupport.SpyOnCalls(&witCallsCounter)).
+			Reply(200)
+		// call to Tenant Service
+		tenantCallsCounter := 0
+		gock.New("http://localhost:8090").
+			Delete(fmt.Sprintf("/api/tenants/%s", userToDeactivate.IdentityID().String())).
+			MatchHeader("Authorization", "Bearer "+saToken).
+			// MatchHeader("X-Request-Id", reqID).
```
Comment (alexeykazakov) [inline authentication/account/service/user_blackbox_test.go]: Same question here.

File: authentication/account/service/user.go (hunk 1 above)
Comment (xcoulon) [reply authentication/account/service/user.go]: right, I initially put the calls to WIT and Tenant in the transaction, so I needed to change this line as well, but I can revert it, now that the calls to other services is out of the transaction scope.

File: authentication/account/service/user.go (hunk 2 above)
Comment (xcoulon) [reply authentication/account/service/user.go]: ok, let me just log the errors but not block. Maybe even for `tenant` BTW.

File: authentication/account/service/user_blackbox_test.go (hunk 3 above)
Comment (xcoulon) [reply authentication/account/service/user_blackbox_test.go]: there was no valid reason 😂

File: authentication/account/service/user_blackbox_test.go (hunk 4 above)
Comment (xcoulon) [reply authentication/account/service/user_blackbox_test.go]: fixed in c7d24bd

File: authentication/account/service/user_blackbox_test.go (hunk 3 above)
Comment (xcoulon) [reply authentication/account/service/user_blackbox_test.go]: fixed in c7d24bd

File: authentication/account/service/user.go (hunk 2 above)
Comment (xcoulon) [reply authentication/account/service/user.go]: fixed in c7d24bd

File: authentication/account/service/user.go (hunk 1 above)
Comment (xcoulon) [reply authentication/account/service/user.go]: fixed in c7d24bd

File: authentication/account/service/user.go (hunk 2 above)
Comment (alexeykazakov) [reply authentication/account/service/user.go]: No. Let’s keep tenant call intact. If we failed to delete namespaces we actually failed to deactivate.

**Comment by xcoulon** [comment]
@alexeykazakov I addressed all your comments. Please let me know if it's ok for you

**Comment by alexeykazakov** [comment]
Just bring back the tenant call error thingy and everything else looks good! :)

File: authentication/account/service/user_blackbox_test.go (hunk 4 above)
Comment (xcoulon) [reply authentication/account/service/user_blackbox_test.go]: > Just bring back the tenant call error thingy and everything else looks good! :)

ok

**Comment by xcoulon** [comment]
> Just bring back the tenant call error thingy and everything else looks good! :)

done in 249cd96

**Pull request merged**